libyul/backends/evm/SSAControlFlowGraph.cpp

```diff
@@ -0,0 +1,249 @@
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/backends/evm/SSAControlFlowGraph.h>
+
+#include <libsolutil/StringUtils.h>
+#include <libsolutil/Visitor.h>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wtautological-compare"
+#include <fmt/ranges.h>
+#pragma GCC diagnostic pop
+
+#include <range/v3/view/zip.hpp>
+
+using namespace solidity;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+namespace
+{
+class SSACFGPrinter
+{
+public:
+	SSACFGPrinter(SSACFG const& _cfg, SSACFG::BlockId _blockId)
+	{
+		printBlock(_cfg, _blockId, 0);
+	}
+	SSACFGPrinter(SSACFG const& _cfg, size_t _functionIndex, Scope::Function const& _function)
+	{
+		printFunction(_cfg, _functionIndex, _function);
+	}
+	friend std::ostream& operator<<(std::ostream& stream, SSACFGPrinter const& printer) {
+		stream << printer.m_result.str();
+		return stream;
+	}
+private:
+	static std::string varToString(SSACFG const& _cfg, SSACFG::ValueId _var) {
+		if (_var.value == std::numeric_limits<size_t>::max())
+			return "INVALID";
+		auto const& info = _cfg.valueInfo(_var);
+		return std::visit(
+			GenericVisitor{
+				[&](SSACFG::UnreachableValue const&) -> std::string {
+					return "[unreachable]";
+				},
+				[&](SSACFG::PhiValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::VariableValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::LiteralValue const& _literal) -> std::string {
+					return formatNumberReadable(_literal.value);
+				}
+			},
+			info
+		);
+	}
+
+	static std::string formatPhi(SSACFG const& _cfg, SSACFG::PhiValue const& _phiValue)
+	{
+		auto const transform = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
+		std::vector<std::string> formattedArgs;
+		formattedArgs.reserve(_phiValue.arguments.size());
+		for (auto const& [arg, entry]: ranges::zip_view(_phiValue.arguments | ranges::views::transform(transform), _cfg.block(_phiValue.block).entries))
+		{
+			formattedArgs.push_back(fmt::format("Block {} => {}", entry.value, arg));
+		}
+		if (!formattedArgs.empty())
+			return fmt::format("φ(\\l\\\n\t{}\\l\\\n)", fmt::join(formattedArgs, ",\\l\\\n\t"));
+		else
+			return "φ()";
+	}
+
+	void writeBlock(SSACFG const& _cfg, SSACFG::BlockId const& _id, SSACFG::BasicBlock const& _block, size_t const _functionIndex)
+	{
+		auto const transform = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
```

Comment:
Maybe `toString` is a better name here?
```suggestion
		auto const toString = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
```